### PR TITLE
refactor(ast): use struct, and enum tuple

### PR DESCRIPTION
**Pull request by yuanbohan**

### CODE REVIEW

File: src/parser/ast.rs [hunk 1]
```diff
@@ -36,104 +36,136 @@ pub struct EvalStmt {
...
+}
+
+#[derive(Debug, Clone)]
+pub struct Unary {
+    pub op: TokenType,
+    pub expr: Box<Expr>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Binary {
```
Comment (evenyag) [inline src/parser/ast.rs:55]: Do we need to keep the name as same as [Prometheus](https://github.com/prometheus/prometheus/blob/0372e259ba/promql/parser/ast.go#L101)

Binary sounds like a binary string.
```suggestion
pub struct BinaryExpr {
```

Comment (yuanbohan) [reply src/parser/ast.rs:55]: I referred [LogicalPlan](https://docs.rs/datafusion-expr/latest/datafusion_expr/enum.LogicalPlan.html), and the Enum name `Expr` can tell us that it is an expression, so I omitted the duplicated suffix `Expr` (and I also copied the idea from go convention).

File: src/parser/ast.rs [hunk 2]
```diff
@@ -36,104 +36,136 @@ pub struct EvalStmt {
     pub lookback_delta: Duration,
 }
 
+#[derive(Debug, Clone)]
+pub struct Aggregate {
+    pub op: TokenType,         // The used aggregation operation.
```
Comment (waynexia) [inline src/parser/ast.rs:41]: Do you think changing this to `op: Token` would make it easier to use or not? `TokenType` is an alias of `u8`, it looks like this https://github.com/GreptimeTeam/greptimedb/pull/870/files#diff-4ad2ad1c2791cacb7e2adcb46c3d68761ab483656a4279213d11960934fd1528R476-R490. And `Token` can do `match` in string literal. Not sure which is better.

**Review by waynexia — COMMENTED** [review]
Mostly looks good to me

File: src/parser/ast.rs [hunk 3]
```diff
@@ -36,104 +36,136 @@ pub struct EvalStmt {
...
+    pub offset: Duration,
+    pub timestamp: Option<i64>,
+    pub start_or_end: TokenType, // Set when @ is used with start() or end()
+    pub step: Duration,
+}
+
+#[derive(Debug, Clone)]
+pub struct NumberLiteral {
+    pub val: f64,
+    pub span: Span,
```
[the remark this answers is not in the document]
Comment (yuanbohan) [reply src/parser/ast.rs:86]: I think so, I just literally copied from [Prometheus](https://github.com/prometheus/prometheus/blob/main/promql/parser/ast.go#L153), and I think it is not useful in execution, the better place for span information fields will be in the lexer

File: src/parser/ast.rs (hunk 2 above)
Comment (yuanbohan) [reply src/parser/ast.rs:41]: string literal is designed for display, and it is better to use the const TokenType, and I think it is more efficient than string in terms of matching.

File: src/parser/ast.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (yuanbohan) [reply src/parser/ast.rs]: I have omitted the span fields

File: src/parser/ast.rs (hunk 2 above)
Comment (waynexia) [reply src/parser/ast.rs:40]: 👍 good

**Review by waynexia — APPROVED** [review]

File: src/parser/ast.rs (hunk 1 above)
Comment (waynexia) [reply src/parser/ast.rs:54]: DataFusion has two `Aggregator`, one for enum aggregator ops and one for aggregate expr struct (like here). It's a bit annoying when I need to refer to both of them, I have to rename one. So I'm +1 to postfix these structs.

Comment (yuanbohan) [reply src/parser/ast.rs:54]: And it convinces me to add the suffix

**Review by evenyag — APPROVED** [review]

**Pull request merged**